py3status/modules/imap_idle.py

```diff
@@ -0,0 +1,152 @@
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    allow_urgent = False
+    criterion = 'UNSEEN'
+    format = 'Mail: {unseen}'
+    hide_if_zero = False
+    mailbox = 'INBOX'
+    password = None
+    port = '993'
+    security = 'ssl'
+    server = None
+    user = None
+
+    class Meta:
+        deprecated = {
+            'rename': [
+                {
+                    'param': 'new_mail_color',
+                    'new': 'color_new_mail',
+                    'msg': 'obsolete parameter use `color_new_mail`',
+                },
+                {
+                    'param': 'imap_server',
+                    'new': 'server',
+                    'msg': 'obsolete parameter use `server`',
+                },
+            ],
+        }
+
+    def post_config_hook(self):
+        self.mail_count = 0 # will be used to store
+        if self.security not in ["ssl", "starttls"]:
+            raise ValueError("Unknown security protocol")
+        self.idle_thread = self.Idle_Thread(
+            mailbox=self.mailbox,
+            criterion=self.criterion,
+            user=self.user,
+            password=self.password,
+            server=self.server,
+            port=self.port,
+            security=self.security)
+        self.idle_thread.start()
+
+    class Idle_Thread(Thread):
+        def __init__(self, mailbox, criterion, user, password, server, port, security):
+            Thread.__init__(self)
+            self.connection = None
+            self.mailbox = mailbox
+            self.criterion = criterion
+            self.user = user
+            self.password = password
+            self.server = server
+            self.port = int(port)
+            self.security = security
+            self.mail_count = 0
+        def get_mail_count(self):
+            return self.mail_count
+        def run(self):
+            while True:
+                try:
+                    if self.connection is None:
+                        if self.security == "ssl":
+                            self.connection = imaplib2.IMAP4_SSL(self.server, self.port)
+                        elif self.security == "starttls":
+                            self.connection = imaplib2.IMAP4(self.server, self.port)
+                            self.connection.starttls(create_default_context())
+                        self.connection.login(self.user, self.password)
+
+                    self.mail_count = 0
+                    directories = self.mailbox.split(',')
+
+                    for directory in directories:
+                        self.connection.select(directory)
+                        unseen_response = self.connection.search(None, self.criterion)
+                        mails = unseen_response[1][0].split()
+                        self.mail_count += len(mails)
+
+                    self.connection.idle()
+                except Exception as e:
+                    import time
+                    time.sleep(1)
```

Comment:
Explain the `time` delay?

Reply:
when an exception was thrown while connecting (e.g. wrong password) the program tried to immediately retry, multiple times per second over and over again. a remnant of bodging; now removed and the while loop moved inside the try block

Reply:
I think... For wrong password, we should throw `self.py3.error` when we can... Also, with `mpd_status` where I threw an error, but `mpd_status` would keep connecting with the sockets? Issue got filed here: https://github.com/ultrabug/py3status/issues/1115. Otherwise, this would keep going just like `mpd_status`'s sockets. For other exception, maybe we don't need `import time`. Anyhow, good to know. Are you finished merging everything over?

Running `imap.py` in the terminal shouldn't give me `Mail: 0`.

Reply:
throwing errors is definitely a good idea, will do (tomorrow). 

I'm having trouble understanding what you mean with #1115 - you mean set cache_timeout to forever when we encounter an error? that seems reasonable. 

import time is gone already, everything has been merged. 

I'll have to check that (again, tomorrow).

Reply:
Wrong issue. #1114. Sorry. Cheers. Sure on tomorrow.

Reply:
if 1114 would get implemented, i wouldn't have to do anything here anyways, correct?      
(getting late here in gmt+1/cet...)

Reply:
I think it would be something like...
```
try:
except ExceptionBadPassword:
   self.py3.error('Bad Password')
except:
   ... close/logout/None
```
Okay. Goodnight. Almost 6pm here.